Improve macOS notarization tools

freakboy3742 opened the pull request:
Makes a number of improvements to Briefcase's wrapper around app notarization for macOS.

1. Replaces the use of `shutil.make_archive` with `ditto`. There is an edge case of app notarization where UTF-8 encoding of some file system metadata can result in notarization failures; this is because the default zip tools don't guarantee that these file system metadata are preserved correctly in the archive. Apple provides a tool called "ditto" that produces zip archives, but guarantees the integrity of file system metadata. The use of ditto is strongly recommended by Apple anyway, so this is broadly a good idea.
2. The time required to notarize an app is highly variable - it can take *hours*, depending on the load at Apple's end. The current notarization tooling requires users to leave a terminal open, and in a position to continuously contact Apple's servers, which is less than ideal. This PR submits the app for notarization, then outputs a submission ID, before going into a polling loop checking for notarization approval. If this loop is interrupted, the user can provide the submission ID as a command line argument to resume the previous notarization pass.
3. If notarisation fails, Briefcase will now output a summary of the notarization problems that have been reported.
4. The filename used when submitting a bare app for notarization now uses the app's formal name. This is mostly for the benefit of debugging when notarization history is inspected - the current name of "archive.zip" doesn't tell you which app has been archived.
5. Some of the helper messages related to selecting identities have been clarified, ensuring that they accurately reflect the context in which they were invoked
6. If verbosity is set to level 3 (i.e., `-vvv`), any multi-threaded tasks are reduced to a single thread. This makes it easier to identify failures when they occur in the multi-threaded section of code (such as app signing).

Fixes #1218 
Fixes #1472

## PR Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] All new features have been tested
- [x] All new features have been documented
- [x] I have read the **CONTRIBUTING.md** file
- [x] I will abide by the code of conduct

mhsmith commented on src/briefcase/platforms/macOS/__init__.py:
```diff
@@ -755,28 +772,129 @@ def verify_app(self, app):
+        if installer_identity:
+            identity_args = (
+                f"--identity {identity.id} "
+                f"--installer_identity {installer_identity.id}"
```
```suggestion
                f"--installer-identity {installer_identity.id}"
```

mhsmith commented on src/briefcase/platforms/macOS/__init__.py:
```diff
@@ -840,19 +957,21 @@ def notarize(self, filename, identity: SigningIdentity):
-                        check=True,
-                    )
-                    submitted = True
+                    with self.input.wait_bar("Submitting app for notariztion..."):
```
```suggestion
                    with self.input.wait_bar("Submitting app for notarization..."):
```

freakboy3742 commented:
> I tried notarizing all 3 formats, but they all failed in different ways.
> As discussed in person, this was fixed by upgrading dmgbuild from 1.6.2 to 1.6.4. Notarization then succeeded, both with and without an interruption.

Corrected by bumping the minimum version requirement from `dmgbuild`.

> .pkg returned some signature errors:

I swear I tested this, but clearly I didn't - I must have checked that it started the notarization wait process, then killed the check and not waited for completion, assuming that if it works for `dmg`, it musts work for `pkg` as well.

It turns out the individual signatures were fine - but when we build the `pkg` file, we duplicate the app into a new location using `shutil.copytree()` ... which unrolls symlinks by default. As a result, when the symlinks are unrolled, the *overall* signature for frameworks (and then the binary, and then the app) fail because the file manifest has changed since it was signed.

This wasn't an issue prior to Briefcase 0.3.20, as the support package didn't contain any symlinks. The most recent support packages use dynamic linking and a repackaged version of the official macOS distribution, so there are symlinks in the Framework package.

> .zip completed successfully when not interrupted, but failed to resume after an interruption:

This was an edge case that didn't get caught because I was doing lots of repeated testing, so I end up in a situation where the file it was looking for *did* exist as a result of a previous testing run. However, this pointed out a bunch of edge cases caused by the fact that .zip distribution artefacts aren't created until *after* notarisation is complete (since it's the .app that is stapled); I've reworked things to address this.
